crates/services/p2p/src/p2p_service.rs

```diff
@@ -635,6 +679,180 @@ mod tests {
+    // It should win every time, that is, there is always one slot reserved for it by the `PeerManager`.
+    #[tokio::test]
+    #[instrument]
+    async fn reserved_nodes_reconnect_works() {
+        let mut p2p_config =
+            Config::default_initialized("reserved_nodes_reconnect_works");
+        // enable mdns for faster discovery of nodes
+        p2p_config.enable_mdns = true;
+
+        // total amount will be `max_peers_allowed` + `reserved_nodes.len()`
+        let max_peers_allowed = 3;
+
+        let bootstrap_nodes_data: Vec<NodeData> = (0..max_peers_allowed * 5)
+            .map(|_| NodeData::random())
+            .collect();
+
+        let reserved_nodes_data: Vec<NodeData> =
+            (0..3).map(|_| NodeData::random()).collect();
+
+        let mut sentry_node = {
+            let mut p2p_config = p2p_config.clone();
+            p2p_config.max_peers_connected = max_peers_allowed as u32;
+
+            p2p_config.bootstrap_nodes = bootstrap_nodes_data
+                .iter()
+                .map(|node| node.multiaddr.clone())
+                .collect();
+
+            p2p_config.reserved_nodes = reserved_nodes_data
+                .iter()
+                .map(|node| node.multiaddr.clone())
+                .collect();
+
+            NodeData::random().create_service(p2p_config)
+        };
+
+        let reserved_node = reserved_nodes_data.last().unwrap().clone();
+        let reserved_node_peer_id =
+            PeerId::from_public_key(&reserved_node.keypair.public());
+
+        let mut all_node_services: Vec<_> = bootstrap_nodes_data
+            .iter()
+            .chain(reserved_nodes_data.iter())
+            .map(|node| node.create_service(p2p_config.clone()))
+            .collect();
+
+        let mut previously_connected = false;
+
+        loop {
+            tokio::select! {
+                sentry_node_event = sentry_node.next_event() => {
+                    if let Some(FuelP2PEvent::PeerConnected(peer_id)) = sentry_node_event {
+                        // we connected to the desired reserved node
+                        if peer_id == reserved_node_peer_id {
+                            // 0. This is the initial connection with the reserved node
+                            if !previously_connected {
+                                // 1. Disconnect and ban the reserved node
+                                // this is done so that other nodes in the network try to take the slot of the reserved node
+                                let _ = sentry_node.swarm.disconnect_peer_id(reserved_node_peer_id);
+                                sentry_node.swarm.ban_peer_id(reserved_node_peer_id);
+                                previously_connected = true;
+                            } else {
+                                // 3. We have successfully re-connected.
+                                // Multiple other nodes in the meantime have tried to establish the connection
```

Comment:
It seems like there could be lots of race conditions that don't exercise this behavior. For instance, the node may unban and reconnect to the reserved node before the other nodes get a chance to connect. Seems like it would be simpler to only start the reserved node once the we've connected to the other peers first.

Reply:
yes that would be even better approach, I will update it